### PR TITLE
Create DB session provider(based on xorm)

### LINKED ISSUE #7137: Gitea can't automatically create table for session storage in MySQL database
- Gitea version (or commit ref): 1.8.2
- Git version:
- Operating system:
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant

Greetings!

I'm using Gitea v.1.8.2 in my setup and have recently ran into strange behaviour: when changing session storage from files to MySQL Gitea doesn't create the needed table in the database. All other tables are being created normally, though (if database was initially empty). Here are some errors from logs, that could be seen right after changing session storage in clean installation from file to MySQL and restarting:

```
May 23 22:09:31 testing gitea[3984]: [Macaron] 2019-05-23 22:09:31: Started GET /metrics for 123.123.123.123
May 23 22:09:31 testing gitea[3984]: [Macaron] PANIC: session/mysql: error checking existence: Error 1146: Table 'testing.session' doesn't exist
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/session/mysql/mysql.go:152 (0xa4cdc3)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/modules/session/virtual.go:67 (0xb3563a)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/session/session.go:290 (0x9e70c0)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/session/session.go:160 (0x9e8379)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:79 (0x9c8580)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x9998e0)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x99960a)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x9c86d9)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:112 (0x9da13f)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/recovery.go:161 (0x9da12d)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:40 (0x9cc373)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x9998e0)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x99960a)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x9c86d9)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:112 (0x9d9460)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:52 (0x9d944b)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:40 (0x9cc373)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x9998e0)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x99960a)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x9c86d9)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/router.go:187 (0x9db386)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/router.go:294 (0x9d4e9d)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/macaron.go:220 (0x9cd74c)
May 23 22:09:31 testing gitea[3984]: /go/src/code.gitea.io/gitea/vendor/github.com/gorilla/context/context.go:141 (0xcdbd9a)
May 23 22:09:31 testing gitea[3984]: /usr/local/go/src/net/http/server.go:1995 (0x6ef443)
May 23 22:09:31 testing gitea[3984]: /usr/local/go/src/net/http/server.go:2774 (0x6f2717)
May 23 22:09:31 testing gitea[3984]: /usr/local/go/src/net/http/server.go:1878 (0x6ee400)
May 23 22:09:31 testing gitea[3984]: /usr/local/go/src/runtime/asm_amd64.s:1337 (0x4620e0)
May 23 22:09:31 testing gitea[3984]: [Macaron] 2019-05-23 22:09:31: Completed GET /metrics 500 Internal Server Error in 833.331�s
```

It's hard to find anywhere the right table structure needed for session table. As workaround I'm creating the table manually on each of mine Gitea instances (using the query from documentation of go-macaron framework):
```
CREATE TABLE IF NOT EXISTS `database_name`.`session` ( `key` CHAR(16) NOT NULL, `data` BLOB, `expiry` INT(11) UNSIGNED NOT NULL, PRIMARY KEY (`key`) ) ENGINE=InnoDB DEFAULT CHARSET=utf8;"
```
Is this a normal behaviour? If not - can u please fix this in future releases (considering that there are no problems with creating all other tables automatically)?

### Steps to reproduce

1. There should be no session table in the database being used
2. Switch session storage to MySQL in Gitea cfg (app.ini), for example:

```
PROVIDER        = mysql
PROVIDER_CONFIG = testing:somepassword@tcp(srv-mysql:3306)/testing
```
3. Restart Gitea for changes to take effect

### PR DESCRIPTION
**Pull request by zeripath**
This PR creates a Xorm session provider which creates
the appropriate Session table for macaron/session.

Fix #7137

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by lafriks** [comment]
lint failure ;)

File: models/migrations/v153.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
```
Comment (6543) [inline models/migrations/v153.go]:
```suggestion
	"code.gitea.io/gitea/modules/timeutil"

```

**Comment by lafriks** [comment]
conflicts ;)

**Comment by zeripath** [comment]
done

File: modules/session/xorm.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+package session
+
+import (
+	"log"
+	"sync"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/timeutil"
+	"gitea.com/macaron/session"
```
Comment (techknowlogick) [inline modules/session/xorm.go]:
```suggestion

	"gitea.com/macaron/session"
```

**Comment by lunny** [comment]
But this still don't resolve the mysql session provider unavailable problem? Could we change mysql as an alias of xorm provider?

**Comment by zeripath** [comment]
well the idea would be that the other db session providers would be considered deprecated.

**Comment by lafriks** [comment]
Imho we should just remove them and add breaking change here. Also we could make it a default one.
Would it be possible to add user id column somehow?

**Comment by zeripath** [comment]
> Imho we should just remove them and add breaking change here.

So removing the mysql and postgres providers may be a little aggressive - happy to remove them if wanted though.

> Also we could make it a default one.

In terms of making it the default I think there is still a place for the filesystem based sessions.

> Would it be possible to add user id column somehow?

Adding the User ID would be a breaking change from the macaron session provider stuff - we'd have to change the API considerably.

**Comment by lafriks** [comment]
I don't see much need for options duplicating functionality.
If no user id can be added than probably type should be renamed to `SessionData` or something like that to keep `Session` reserved for actual user session management in future

**Comment by zeripath** [comment]
The tablename `session` is required to maintain compatibility with the original mysql/postgres provider tables.

**Comment by lafriks** [comment]
I don't think thats much of the problem that we lose session data, it's not critical

**Comment by zeripath** [comment]
conflicts fixed

**Comment by lunny** [comment]
Since macaron will be replaced by chi, I would like to merge this after macaron has been replaced.

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by zeripath** [comment]
@lunny I've updated.

I don't think there's any changes with chi-session but you might be more aware of these.

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -286,6 +286,8 @@ var migrations = []Migration{
 	NewMigration("Recreate user table to fix default values", recreateUserTableToFixDefaultValues),
 	// v169 -> v170
 	NewMigration("Update DeleteBranch comments to set the old_ref to the commit_sha", commentTypeDeleteBranchUseOldRef),
+	// v170 -> v171
+	NewMigration("Add sessions table for macaron/session", addSessionTable),
```
Comment (lunny) [inline models/migrations/migrations.go]: Now the name changed to `go-chi/session`

File: models/session.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+// Session represents a session compatible for go-macaron session
+type Session struct {
+	Key    string             `xorm:"pk CHAR(16)"` // has to be Key to match with go-macaron/session
```
Comment (lunny) [inline models/session.go]: go-chi/session

File: models/session.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	Data   []byte             `xorm:"BLOB"`
+	Expiry timeutil.TimeStamp // has to be Expiry to match with go-macaron/session
+}
+
+// UpdateSession updates the session with provided id
+func UpdateSession(key string, data []byte) error {
+	_, err := x.Update(&Session{
```
Comment (lunny) [inline models/session.go]:
```go
x.ID(key).Update(&Session{
    Data:   data,
		Expiry: timeutil.TimeStampNow(),
})
```

File: models/session.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	if _, err := sess.Exec("UPDATE "+sess.Engine().TableName(&Session{})+" SET `key` = ? WHERE `key`=?", newKey, oldKey); err != nil {
+		return nil, err
+	}
+
+	s := Session{
+		Key: newKey,
+	}
+	if _, err := sess.Get(s); err != nil {
```
Comment (lunny) [inline models/session.go]:
```go
if _, err := sess.Get(&s); err != nil {
```

**Review by lunny — CHANGES_REQUESTED** [review]
And I would like name it `DB` session.

File: models/migrations/migrations.go (hunk 3 above)
Comment (zeripath) [reply models/migrations/migrations.go]:
```suggestion
	NewMigration("Add sessions table for go-chi/session", addSessionTable),
```

File: models/session.go (hunk 4 above)
Comment (zeripath) [reply models/session.go]:
```suggestion
	Key    string             `xorm:"pk CHAR(16)"` // has to be Key to match with go-chi/session
```

File: models/session.go (hunk 5 above)
Comment (zeripath) [reply models/session.go]:
```suggestion
	_, err := x.ID(key).Update(&Session{
		Data:   data,
		Expiry: timeutil.TimeStampNow(),
	})
```

File: models/session.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	Expiry timeutil.TimeStamp // has to be Expiry to match with go-macaron/session
+}
+
+// UpdateSession updates the session with provided id
+func UpdateSession(key string, data []byte) error {
+	_, err := x.Update(&Session{
+		Key:    key,
+		Data:   data,
+		Expiry: timeutil.TimeStampNow(),
+	})
```
Comment (zeripath) [inline models/session.go]:
```suggestion
```

File: models/session.go (hunk 6 above)
Comment (zeripath) [reply models/session.go]:
```suggestion
	if _, err := sess.Get(&s); err != nil {
```

**Comment by zeripath** [comment]
You mean from XormSession to DBSession?

**Review by 6543 — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by techknowlogick** [comment]
🚀

**Pull request merged**